dist_ir/backend/torch.py

```diff
@@ -446,42 +545,163 @@ def run_process(ctx, num_warmup_steps, num_repetitions, rank, fn, inputs):
+                # Before starting the op runtime profiling, we sychronize all
+                # devices to make sure all op runtimes begin from the same point.
+                if i == num_warmup_steps:
+                    if ctx.world_size > 0:
+                        torch.distributed.barrier()
+                    elif ctx.use_gpu:
+                        torch.cuda.synchronize(device=rank)
+                record_op_runtimes = i >= num_warmup_steps
+
+            print(f"Rank {rank}: Dispatching step {i}...")
+
+            # We add a timing event at the start of every iteration, and at the
+            # end of the last iteration. This ensures that the delta between any
+            # two consecutive timing events i and i + 1 measures the runtime of
+            # iteration i.
+            #
+            # Note that we do not need any synchronization between these events
+            # because we compute the latency for a given iteration by taking the
+            # maximum runtime observed across all ranks in that iteration.
+            add_event(ctx, events)
+            if i == 0:
+                # We run the first iteration with a try/catch block to gracefully
+                # exit in the event of errors. After confirming there are no errors,
+                # we proceed without the try/catch in subsequent iterations.
+                try:
+                    outputs, peak_memory, op_runtime_events = run_function(
+                        ctx,
+                        fn,
+                        inputs,
+                        rank,
+                        recv_buffers,
+                        record_op_runtimes=record_op_runtimes,
+                    )
+                except Exception as e:
+                    print_exc()
+                    return None, None, None
+            else:
+                outputs, peak_memory, op_runtime_events = run_function(
+                    ctx,
+                    fn,
+                    inputs,
+                    rank,
+                    recv_buffers,
+                    record_op_runtimes=record_op_runtimes,
+                )
+            if i == (num_warmup_steps + num_repetitions - 1):
+                add_event(ctx, events)
+
+            if record_op_runtimes:
+                # We synchronize here to make sure all computation has completed
+                # before measuring op durations.
+                if ctx.world_size > 1:
+                    torch.distributed.barrier()
+                elif ctx.use_gpu:
+                    torch.cuda.synchronize(device=rank)
+
+                # We measure the duration of each op as the delta between each pair
+                # of events i and j where i is even and j = i + 1. Note that this
+                # differs from how we measure durations for the entire iteration,
+                # as in this case we want to ignore any overhead from de-allocating
+                # memory when constructing the trace.
```

Comment:
The de-allocation happening here: https://github.com/microsoft/dist-ir/blob/memory_tracking/dist_ir/backend/torch.py#L508
That's independent of the actual op execution time but still happens every time we execute an op (notice how the op runtime event is recorded earlier at line 486)
We may want to separately measure the de-allocation time and include that in the trace as well but so far I haven't done that yet

Reply:
I see, thanks. It would be helpful if you could update this comment to point to the deallocation, and maybe add a TODO to include deallocation in the trace.